File: day4/src/main.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use std::collections::HashSet;

fn part1(filename: &str) -> io::Result<()> {
    let path = Path::new(filename);
    let file = File::open(&path)?;
    let reader = io::BufReader::new(file);

    let mut ans = 0;

    for line in reader.lines() {
        let line = line.unwrap();

        let (_, numbers) = line.split_once(": ").unwrap();
        let (winning_numbers, our_numbers) = numbers.split_once(" | ").unwrap();
        let mut exponent = -1;
        let winning: HashSet<i32> = winning_numbers.split_whitespace().filter_map(|x| {
            x.parse::<i32>().ok()
        }).collect();

        for number in our_numbers.split(" ") {
            let num = number.parse::<i32>();
            match num {
                Ok(num) => {
                    if winning.contains(&num) {
                        exponent += 1
                    }
                },
                Err(_) => {}
            }
        }

        if exponent >= 0 {
            ans += 2_i32.pow(exponent as u32);
        }
        
    }

    println!("Part 1: {}", ans);

    Ok(())
}

fn part2(filename: &str) -> io::Result<()> {
    let path = Path::new(filename);
    let file = File::open(&path)?;
    let reader = io::BufReader::new(file);

    let mut scratchcards = vec![1; 300];
    scratchcards[0] = 0;

    let mut last_card_id = 0;

    for line in reader.lines() {
        let line = line.unwrap();

        let (card, numbers) = line.split_once(": ").unwrap();

        let card_id = card.split_whitespace().last().unwrap().parse::<usize>().unwrap();
        last_card_id = card_id;
        if scratchcards[card_id] == 0 {
            continue;
        }

        let (winning_numbers, our_numbers) = numbers.split_once(" | ").unwrap();
        let mut won = 0;
        let winning: HashSet<i32> = winning_numbers.split_whitespace().filter_map(|x| {
            x.parse::<i32>().ok()
        }).collect();

        for number in our_numbers.split(" ") {
            let num = number.parse::<i32>();
            match num {
                Ok(num) => {
                    if winning.contains(&num) {
                        won += 1
                    }
                },
                Err(_) => {}
            }
        }

        for i in card_id+1..card_id+won+1 {
            scratchcards[i] += scratchcards[card_id];
        }
        
    }

    for i in last_card_id+1..scratchcards.len() {
        scratchcards[i] = 0;
    }

    let ans: i32 = scratchcards.iter().sum();

    println!("Part 2: {}", ans);

    Ok(())
}

fn main() {
    part1("./src/input.txt").unwrap();
    part2("./src/input.txt").unwrap();
}

File: day4/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func part1(filename string) {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	ans := 0

	scanner := bufio.NewScanner(file)
	lines := []string{}

	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	bingoNumbers := []int{}

	for _, number := range strings.Split(lines[0], ",") {
		iNumber, _ := strconv.Atoi(number)
		bingoNumbers = append(bingoNumbers, iNumber)
	}

	type bingo [5][5]int

	bingos := []bingo{}

	splitFn := func(c rune) bool {
		return c == ' '
	}

	for i := 2; i < len(lines); i += 6 {
		b := bingo{}
		for j := 0; j < 5; j++ {
			for k, number := range strings.FieldsFunc(lines[i+j], splitFn) {
				iNumber, _ := strconv.Atoi(number)
				b[j][k] = iNumber
			}
		}
		bingos = append(bingos, b)
	}

	markBingo := func(bingo *bingo, number int) {
		for i := 0; i < 5; i++ {
			for j := 0; j < 5; j++ {
				if bingo[i][j] == number {
					bingo[i][j] = -1
				}
			}
		}
	}

	checkBingo := func(bingo bingo) bool {
		for i := 0; i < 5; i++ {
			good := true

			for j := 0; j < 5; j++ {
				if bingo[i][j] != -1 {
					good = false
					break
				}
			}

			if good {
				return true
			}
		}

		for j := 0; j < 5; j++ {
			good := true

			for i := 0; i < 5; i++ {
				if bingo[i][j] != -1 {
					good = false
					break
				}
			}

			if good {
				return true
			}
		}

		return false
	}

	sumBingo := func(bingo bingo) int {
		sum := 0

		for i := 0; i < 5; i++ {
			for j := 0; j < 5; j++ {
				if bingo[i][j] != -1 {
					sum += bingo[i][j]
				}
			}
		}

		return sum
	}

out:
	for _, nr := range bingoNumbers {
		for i := 0; i < len(bingos); i++ {
			markBingo(&bingos[i], nr)
			if checkBingo(bingos[i]) {
				ans = sumBingo(bingos[i]) * nr
				break out
			}
		}
	}

	fmt.Println("Part 1: ", ans)
}

func part2(filename string) {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	ans := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
	}

	fmt.Println("Part 1: ", ans)
}

func main() {
	part1("./input.txt")
	part2("./input.txt")
}
